svf/include/Graphs/CallGraph.h

```diff
@@ -0,0 +1,345 @@
+//===- CallGraph.h -- Call graph representation----------------------------//
+//
+//                     SVF: Static Value-Flow Analysis
+//
+// Copyright (C) <2013-2017>  <Yulei Sui>
+//
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+//===----------------------------------------------------------------------===//
+
+/*
+ * CallGraph.h
+ *
+ *  Created on: Nov 7, 2013
+ *      Author: Yulei Sui
+ */
+
+#ifndef CALLGRAPH_H_
+#define CALLGRAPH_H_
+
+#include "Graphs/GenericGraph.h"
+#include "SVFIR/SVFValue.h"
+#include "Graphs/ICFG.h"
+#include <set>
+
+namespace SVF
+{
+
+class CallGraphNode;
+class SVFModule;
+
+
+/*
+ * Call Graph edge representing a calling relation between two functions
+ * Multiple calls from function A to B are merged into one call edge
+ * Each call edge has a set of direct callsites and a set of indirect callsites
+ */
+typedef GenericEdge<CallGraphNode> GenericCallGraphEdgeTy;
+class CallGraphEdge : public GenericCallGraphEdgeTy
+{
+
+public:
+    typedef Set<const CallICFGNode*> CallInstSet;
+    enum CEDGEK
+    {
+        CallRetEdge,TDForkEdge,TDJoinEdge,HareParForEdge
+    };
+
+
+private:
+    CallInstSet directCalls;
+    CallSiteID csId;
+public:
+    /// Constructor
+    CallGraphEdge(CallGraphNode* s, CallGraphNode* d, CEDGEK kind, CallSiteID cs) :
+        GenericCallGraphEdgeTy(s, d, makeEdgeFlagWithInvokeID(kind, cs)), csId(cs)
+    {
+    }
+    /// Destructor
+    virtual ~CallGraphEdge()
+    {
+    }
+    /// Compute the unique edgeFlag value from edge kind and CallSiteID.
+    static inline GEdgeFlag makeEdgeFlagWithInvokeID(GEdgeKind k, CallSiteID cs)
+    {
+        return (cs << EdgeKindMaskBits) | k;
+    }
+    /// Get direct and indirect calls
+    //@{
+    inline CallSiteID getCallSiteID() const
+    {
+        return csId;
+    }
+    inline CallInstSet& getDirectCalls()
+    {
+        return directCalls;
+    }
+    inline const CallInstSet& getDirectCalls() const
+    {
+        return directCalls;
+    }
+    //@}
+
+    /// Add direct callsite
+    //@{
+    void addDirectCallSite(const CallICFGNode* call);
```

Comment:
can be made protected

Reply:
callgraphedge's addDirectCallSite is used in CallGraph, so it cannot be protected

Reply:
If it is only used in 'CallGraph', it can be made as private.